File: day_02/solution.js
const fs = require('fs')

const readCommands = (filePath) => {
  let commands = [];
  try {
    const fileData = fs.readFileSync(filePath, 'utf8');
    commands = fileData.split(/\r?\n/).map(command => {
        let commandParts = command.split(' ');
        let direction = commandParts[0];
        let distance = parseInt(commandParts[1]);
        return {direction, distance};
    });
  } catch (err) {
    console.error(err);
  }
  return commands;
}

const part1 = commands => {
  let horizontal = 0;
  let depth = 0;
  commands.forEach(command => {
      switch (command.direction) {
        case 'forward':
            horizontal += command.distance;
            break
        case 'up':
            depth -= command.distance;
            break
        case 'down':
            depth += command.distance;
            break
      }; 
  });
  return horizontal * depth;
}

const part2 = commands => {
  let horizontal = 0;
  let depth = 0;
  let aim = 0;
  commands.forEach(command => {
    switch (command.direction) {
      case 'forward':
          horizontal += command.distance;
          depth += command.distance * aim;
          break
      case 'up':
          aim -= command.distance;
          break
      case 'down':
          aim += command.distance;
          break
    }; 
  });
  return horizontal * depth;
}

const main = async () => {
  const testInputData = await readCommands('./day_02/test_input.txt');
  console.log(`Part 1 (test): ${part1(testInputData)}`);
  console.log(`Part 2 (test): ${part2(testInputData)}`);

  const inputData = await readCommands('./day_02/input.txt');
  console.log(`Part 1: ${part1(inputData)}`);
  console.log(`Part 2: ${part2(inputData)}`);
}

main();

/*
const part1 = depths => countSlidingWindowIncrease(depths, 1);
const part2 = depths => countSlidingWindowIncrease(depths, 3);

const main = async () => {
  const testInputData = await readDepths('./day_01/test_input.txt');
  console.log(`Part 1 (test): ${part1(testInputData)}`);
  console.log(`Part 2 (test): ${part2(testInputData)}`);

  const inputData = await readDepths('./day_01/input.txt');
  console.log(`Part 1: ${part1(inputData)}`);
  console.log(`Part 2: ${part2(inputData)}`);
}

main();
*/

File: day_02/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func check(err error) {
	if err != nil {
		panic(err)
	}
}

func toInt(s string) int {
	result, err := strconv.Atoi(s)
	check(err)
	return result
}

type Command struct {
	direction string
	distance  int
}

func readCommands(filename string) []Command {
	file, err := os.Open(filename)
	check(err)
	defer file.Close()

	Scanner := bufio.NewScanner(file)

	var commands []Command
	for Scanner.Scan() {
		s := strings.Split(Scanner.Text(), " ")
		commands = append(commands, Command{s[0], toInt(s[1])})
	}
	return commands
}

func part1(commands []Command) int {
	horizontal := 0
	depth := 0
	for i := 0; i < len(commands); i++ {
		command := commands[i]
		switch command.direction {
		case "forward":
			horizontal += command.distance
		case "up":
			depth -= command.distance
		case "down":
			depth += command.distance
		}
	}
	return horizontal * depth
}

func part2(commands []Command) int {
	horizontal := 0
	depth := 0
	aim := 0
	for i := 0; i < len(commands); i++ {
		command := commands[i]
		switch command.direction {
		case "forward":
			horizontal += command.distance
			depth += command.distance * aim
		case "up":
			aim -= command.distance
		case "down":
			aim += command.distance
		}
	}
	return horizontal * depth
}

func main() {
	testInputData := readCommands("./day_02/test_input.txt")
	fmt.Println("Part 1 (test):", part1(testInputData))
	fmt.Println("Part 2 (test):", part2(testInputData))

	inputData := readCommands("./day_02/input.txt")
	fmt.Println("Part 1:", part1(inputData))
	fmt.Println("Part 2:", part2(inputData))
}
